modules/stochastic_tools/src/utils/CovarianceFunction.C

```diff
@@ -0,0 +1,402 @@
+
+void
+CovarianceKernel::set_noise_variance(const Real sigma_n_squared)
+{
+  _sigma_n_squared = sigma_n_squared;
+}
+
+void
+CovarianceKernel::set_length_factor(const std::vector<Real> length_factor)
+{
+  _length_factor = length_factor;
+}
+
+void
+CovarianceKernel::set_gamma(const Real /*gamma*/)
+{
+  ::mooseError(
+      "Attempting to set gamma hyper parameter for covariance function which does not utilize.");
+}
+
+void
+CovarianceKernel::set_p(const unsigned int /*p*/)
+{
+  ::mooseError(
+      "Attempting to set p hyper parameter for covariance function which does not utilize.");
+}
+
+// begin squared_exponential
+SquaredExponential::SquaredExponential() : CovarianceKernel() {}
+
+RealEigenMatrix
+SquaredExponential::compute_K(const RealEigenMatrix x,
+                              const RealEigenMatrix xp,
+                              const bool is_self_covariance) const
+{
+  unsigned int num_samples_x = x.rows();
+  unsigned int num_samples_xp = xp.rows();
+  unsigned int num_params_x = x.cols();
+  unsigned int num_params_xp = xp.cols();
+
+  if (num_params_x != num_params_xp)
+  {
+    ::mooseError("Number of parameters do not match in covariance kernel calculation");
+  }
+
+  RealEigenMatrix K(num_samples_x, num_samples_xp);
+
+  for (unsigned int ii = 0; ii < num_samples_x; ii++)
+  {
+    for (unsigned int jj = 0; jj < num_samples_xp; jj++)
+    {
+      Real val = 0;
+      for (unsigned int kk = 0; kk < num_params_x; kk++)
+      {
+        // Compute distance per parameter, scaled by length factor
+        val += std::pow((x(ii, kk) - xp(jj, kk)), 2) / (std::pow(_length_factor.at(kk), 2));
+      }
+      val = _sigma_f_squared * std::exp(-val / 2.0);
+      K(ii, jj) = val;
+    }
+    if (is_self_covariance)
+    {
+      K(ii, ii) += _sigma_n_squared;
+    }
+  }
+
+  return K;
+}
+
+void
+SquaredExponential::store(std::ostream & stream, void * context) const
+{
+  std::string type = "Squared Exponential";
+  dataStore(stream, type, context);
+  dataStore(stream, _sigma_f_squared, context);
+  dataStore(stream, _sigma_n_squared, context);
+  unsigned int n = _length_factor.size();
+  dataStore(stream, n, context);
+  for (unsigned int ii = 0; ii < n; ii++)
+  {
+    dataStore(stream, _length_factor.at(ii), context);
+  }
+}
+// end squared_exponential
+
+// begin exponential
+Exponential::Exponential() : CovarianceKernel() {}
+
+RealEigenMatrix
+Exponential::compute_K(const RealEigenMatrix x,
+                       const RealEigenMatrix xp,
+                       const bool is_self_covariance) const
+{
+  unsigned int num_samples_x = x.rows();
+  unsigned int num_samples_xp = xp.rows();
+  unsigned int num_params_x = x.cols();
+  unsigned int num_params_xp = xp.cols();
+
+  if (num_params_x != num_params_xp)
+  {
+    ::mooseError("Number of parameters do not match in covariance kernel calculation");
+  }
+
+  RealEigenMatrix K(num_samples_x, num_samples_xp);
+
+  for (unsigned int ii = 0; ii < num_samples_x; ii++)
```

Comment:
In MOOSE, we have the standard of doing `++ii`. Please change this in all your for loops.